### ISSUE TITLE
[OpenBMC] firmware sorting is not consistent on openbmc objects

### ISSUE THREAD

**Issue by whowutwut** [xcat2/xcat-core#4126]
I noticed this after I flashed mid05tor12cn11 with 1742 firmware, the BMC gets sorted at the bottom for some reason....

See output in: https://github.com/xcat2/xcat-core/pull/4115

Today it's OK
```
[root@briggs01 Packages]# rinv mid05tor12cn11 firm
mid05tor12cn11: BMC Firmware Product: ibm-v1.99.10-113-g65edf7d-r1-0-g84427f0 (Active)*
mid05tor12cn11: HOST Firmware Product: IBM-witherspoon-ibm-OP9_v1.19_1.64 (Active)*
mid05tor12cn11: HOST Firmware Product: -- additional info: buildroot-2017.8-8-g5e23247
mid05tor12cn11: HOST Firmware Product: -- additional info: capp-ucode-p9-dd2-v2
mid05tor12cn11: HOST Firmware Product: -- additional info: hostboot-b1e6bb
mid05tor12cn11: HOST Firmware Product: -- additional info: hostboot-binaries-da1227e
mid05tor12cn11: HOST Firmware Product: -- additional info: linux-4.13.5-openpower1-p25f4095
mid05tor12cn11: HOST Firmware Product: -- additional info: machine-xml-389c56f
mid05tor12cn11: HOST Firmware Product: -- additional info: occ-ba4e81e
mid05tor12cn11: HOST Firmware Product: -- additional info: op-build-v1.7-1472-g9b410c9
mid05tor12cn11: HOST Firmware Product: -- additional info: petitboot-v1.6.0-pc27d9eb
mid05tor12cn11: HOST Firmware Product: -- additional info: sbe-374add4
mid05tor12cn11: HOST Firmware Product: -- additional info: skiboot-v5.9-rc2
```

But I just flashed cn16 to 1742....

```
[root@briggs01 Packages]# rinv mid05tor12cn16 firm
mid05tor12cn16: HOST Firmware Product: IBM-witherspoon-ibm-OP9_v1.19_1.64 (Active)*
mid05tor12cn16: HOST Firmware Product: -- additional info: buildroot-2017.8-8-g5e23247
mid05tor12cn16: HOST Firmware Product: -- additional info: capp-ucode-p9-dd2-v2
mid05tor12cn16: HOST Firmware Product: -- additional info: hostboot-b1e6bb
mid05tor12cn16: HOST Firmware Product: -- additional info: hostboot-binaries-da1227e
mid05tor12cn16: HOST Firmware Product: -- additional info: linux-4.13.5-openpower1-p25f4095
mid05tor12cn16: HOST Firmware Product: -- additional info: machine-xml-389c56f
mid05tor12cn16: HOST Firmware Product: -- additional info: occ-ba4e81e
mid05tor12cn16: HOST Firmware Product: -- additional info: op-build-v1.7-1472-g9b410c9
mid05tor12cn16: HOST Firmware Product: -- additional info: petitboot-v1.6.0-pc27d9eb
mid05tor12cn16: HOST Firmware Product: -- additional info: sbe-374add4
mid05tor12cn16: HOST Firmware Product: -- additional info: skiboot-v5.9-rc2
mid05tor12cn16: BMC Firmware Product: ibm-v1.99.10-113-g65edf7d-r1-0-g84427f0 (Active)*
```

We do put the data into an array and sort this arrray when we print it out, why is this happening? 

The problem is `xcoll` won't work correctly...
```
[root@briggs01 Packages]# rinv mid05tor12cn[11,16] firm | grep -i ibm | xcoll
====================================
mid05tor12cn11
====================================
BMC Firmware Product: ibm-v1.99.10-113-g65edf7d-r1-0-g84427f0 (Active)*
HOST Firmware Product: IBM-witherspoon-ibm-OP9_v1.19_1.64 (Active)*

====================================
mid05tor12cn16
====================================
HOST Firmware Product: IBM-witherspoon-ibm-OP9_v1.19_1.64 (Active)*
BMC Firmware Product: ibm-v1.99.10-113-g65edf7d-r1-0-g84427f0 (Active)*

```

If we run it through sort before `xcoll`.. it is OK

```
[root@briggs01 Packages]# rinv mid05tor12cn[11,16] firm | sort | grep -i ibm | xcoll
====================================
mid05tor12cn16,mid05tor12cn11
====================================
BMC Firmware Product: ibm-v1.99.10-113-g65edf7d-r1-0-g84427f0 (Active)*
HOST Firmware Product: IBM-witherspoon-ibm-OP9_v1.19_1.64 (Active)*

```

Can someone check the sorting algorithm for the sorted_output...?

**Comment by whowutwut** [comment]
@chenglch  If you have some ideas about this, or can you check the sorting code around here: https://github.com/xcat2/xcat-core/blob/1dd8943ea0efe336f6b8bf0438b265f2a46d3592/xCAT-server/lib/xcat/plugins/openbmc.pm#L1667

**Comment by whowutwut** [comment]
I am seeing this on WSC cluster also...

```
[root@c699mgt00 iter-01]# rinv c699c038 firm
c699c038: HOST Firmware Product: IBM-witherspoon-ibm-OP9_v1.19_1.68 (Active)*
c699c038: HOST Firmware Product: -- additional info: buildroot-2017.8-8-g5e23247
c699c038: HOST Firmware Product: -- additional info: capp-ucode-p9-dd2-v2
c699c038: HOST Firmware Product: -- additional info: hostboot-b1e6bb
c699c038: HOST Firmware Product: -- additional info: hostboot-binaries-da1227e
c699c038: HOST Firmware Product: -- additional info: linux-4.13.5-openpower1-pc511380
c699c038: HOST Firmware Product: -- additional info: machine-xml-389c56f
c699c038: HOST Firmware Product: -- additional info: occ-ba4e81e
c699c038: HOST Firmware Product: -- additional info: op-build-v1.7-1473-g6f195a1
c699c038: HOST Firmware Product: -- additional info: petitboot-v1.6.0-p3f89369
c699c038: HOST Firmware Product: -- additional info: sbe-374add4
c699c038: HOST Firmware Product: -- additional info: skiboot-v5.9-rc3
c699c038: BMC Firmware Product: ibm-v1.99.10-113-g65edf7d-r3-0-g9e4f715 (Active)*
```

Why is the BMC sorted after HOST?

**Comment by zet809** [comment]
Hi, @xuweibj , will you pls help to take a look at? Thx!

**Comment by xuweibj** [comment]
Hi @whowutwut , I found that we sort the output with ``$sw_id``, and then rm it before print. Do you know why? 
The comments in our code:
```
 # The firmware output requires the ID to be part of the string to sort correctly.
 # Remove this ID from the output to the user
```
And print out information with ID like this:
```
[30ee1c48]HOST Firmware Product:   IBM-witherspoon-ibm-OP9_v1.19_1.112 (Active)*
[30ee1c48]HOST Firmware Product: -- additional info: op-build-v1.7-1507-g9a24243
[30ee1c48]HOST Firmware Product: -- additional info: buildroot-2017.08-8-g5e23247
[30ee1c48]HOST Firmware Product: -- additional info: skiboot-v5.9-rc4
[30ee1c48]HOST Firmware Product: -- additional info: hostboot-e8c85bb
[30ee1c48]HOST Firmware Product: -- additional info: linux-4.13.7-openpower1-p1c46946
[30ee1c48]HOST Firmware Product: -- additional info: petitboot-v1.6.1-p01d8f7a
[30ee1c48]HOST Firmware Product: -- additional info: machine-xml-eee619a
[30ee1c48]HOST Firmware Product: -- additional info: occ-cff91e5
[30ee1c48]HOST Firmware Product: -- additional info: hostboot-binaries-edd7653
[30ee1c48]HOST Firmware Product: -- additional info: capp-ucode-p9-dd2-v2
[30ee1c48]HOST Firmware Product: -- additional info: sbe-ec9a99e
[30ee1c48]HOST Firmware Product:   IBM-witherspoon-ibm-OP9_v1.19_1.112 (Active)*
[30ee1c48]HOST Firmware Product: -- additional info: op-build-v1.7-1507-g9a24243
[30ee1c48]HOST Firmware Product: -- additional info: buildroot-2017.08-8-g5e23247
[30ee1c48]HOST Firmware Product: -- additional info: skiboot-v5.9-rc4
[30ee1c48]HOST Firmware Product: -- additional info: hostboot-e8c85bb
[30ee1c48]HOST Firmware Product: -- additional info: linux-4.13.7-openpower1-p1c46946
[30ee1c48]HOST Firmware Product: -- additional info: petitboot-v1.6.1-p01d8f7a
[30ee1c48]HOST Firmware Product: -- additional info: machine-xml-eee619a
[30ee1c48]HOST Firmware Product: -- additional info: occ-cff91e5
[30ee1c48]HOST Firmware Product: -- additional info: hostboot-binaries-edd7653
[30ee1c48]HOST Firmware Product: -- additional info: capp-ucode-p9-dd2-v2
[30ee1c48]HOST Firmware Product: -- additional info: sbe-ec9a99e
[f6590ce0]BMC Firmware Product:   ibm-v1.99.10-113-g65edf7d-r8-0-g713d86d (Active)*
mid05tor12cn05: HOST Firmware Product: IBM-witherspoon-ibm-OP9_v1.19_1.112 (Active)*
mid05tor12cn05: HOST Firmware Product: -- additional info: buildroot-2017.8-8-g5e23247
mid05tor12cn05: HOST Firmware Product: -- additional info: capp-ucode-p9-dd2-v2
mid05tor12cn05: HOST Firmware Product: -- additional info: hostboot-binaries-edd7653
mid05tor12cn05: HOST Firmware Product: -- additional info: hostboot-e8c85bb
mid05tor12cn05: HOST Firmware Product: -- additional info: linux-4.13.7-openpower1-p1c46946
mid05tor12cn05: HOST Firmware Product: -- additional info: machine-xml-eee619a
mid05tor12cn05: HOST Firmware Product: -- additional info: occ-cff91e5
mid05tor12cn05: HOST Firmware Product: -- additional info: op-build-v1.7-1507-g9a24243
mid05tor12cn05: HOST Firmware Product: -- additional info: petitboot-v1.6.1-p1d8f7a
mid05tor12cn05: HOST Firmware Product: -- additional info: sbe-ec9a99e
mid05tor12cn05: HOST Firmware Product: -- additional info: skiboot-v5.9-rc4
mid05tor12cn05: BMC Firmware Product: ibm-v1.99.10-113-g65edf7d-r8-0-g713d86d (Active)*
```

If we do not need to print out ID, I think we could not add it.

**Comment by xuweibj** [comment]
I think I got the point why we sort with ``$sw_id``, I will fix this issue today.

**Comment by whowutwut** [comment]
👍 Yea the sorting is so the additional info ends up under the host firmware that it belongs to and the bmc info should be in top

**Comment by xuweibj** [comment]
The issue was fixed in pull request #4266.

**Comment by xuweibj** [comment]
Hi @whowutwut , could this issue be closed?
